content/posts/gsoc-2024-auto-sync-sail/index.md

```diff
@@ -0,0 +1,151 @@
+
+The Capstone project maintains an [update tool](https://rizin.re/posts/auto-sync/) called [Auto-Sync](https://github.com/capstone-engine/capstone/blob/next/suite/auto-sync/intro.md), which can semi-automatically synchronize changes from LLVM to Capstone (using Tree-sitter magic). Alas, it can only do that for some architectures, and RISC-V is not among the supported ones. Also, we already saw how even LLVM is not completely on top of all updates. Fortunately, the solution exists, just hiding elswhere.
+
+
+# To Sail the high seas and RISC it all
+The problem of describing Instruction Set Architectures (ISAs) accurately so that we can do plenty of useful things to them (assembly, disassembly, emulation, codegen, etc…) faces many projects, so much so that some smart people have developed an entirely new special language for it, Sail. Sail is a language designed specifically to address the problem of describing all aspects of ISAs: how the instructions are encoded into binary, how they execute, etc…
+
+Now, if only there was a project that used Sail to describe RISC-V… wait, there is! It’s called Sail-RISCV. It’s such a complete and up-to-date description of RISC-V that the RISC-V foundation adopted it as the official source of truth for the architecture, this means that however the Sail code behaves, is - by definition - how RISC-V should behave.
+
+>> Other architectures modelled in Sail are several versions of [ARM](https://github.com/rems-project/sail-arm/tree/master), a considerable part of [x86](https://github.com/rems-project/sail-x86-from-acl2), and a research version of MIPS called [CHERI-MIPS](https://github.com/CTSRD-CHERI/sail-cheri-mips), which includes hardware extensions to assist and accelerate memory safe pointers. The ARM and x86 models are auto-generated from other descriptions, and all 3 models are much less active than RISC-V's.
+
+Let’s see a snippet of what Sail looks like in practice, here’s the definition of RISC-V IType (immediate) instructions:
+![IMAGE__SAIL_SNIPPET_ITYPE_ENCDEC](https://hackmd.io/_uploads/ryhunMMBkg.png)
+
+The rule might be as cryptic as latin if you’re not used to pattern-matching constructs from functional languages, but what it’s saying is simply the following:
+1. If the first (least-significant) 7 bits of the 32-bit instruction are: 0010011
+1. And if the 3 bits from bit 12 to bit 14 are in the table encdec_iop
+1. Then, the operation specified by this instruction is whatever enum corresponding to bits 12:14 in the encdec_iop table, and the args are:
+    * The 5-bit register index **rd** in bits 7 through 11
+    * The 5-bit register index **rs1** in bits 15 through 19
+    * The 12-bit literal **imm** in bits 20 through 31
+4. Otherwise, if (1) and (2) are not true, keep checking the 32-bit binary instruction against other rules
```

Comment:
Which line does this in he source above?

Reply:
If you mean the destructuring of the instruction into [`rd, rs1, imm`], it's implicit in the way the mapping pattern is constructed (last line in the image), `@` is the vector concatenation operator in Sail. The actual bit widths are inferred because the types of `rd, rs1,` and `imm` were all declared before in the AST clause declaration (first line, `regidx` was declared earlier in the project as a synonym for a bitvector of width 5).

If you mean point (4), that's just pattern matching proceeding sequentially across the clauses of `encdec` until it succeeds. The only difference is that Sail has a unique and innovative feature of "Scattered Mappings", which means you can write different rules for `encdec` across different lines and different source files (separated by other logic and definitions) and Sail will still consider all of them part of the same logical pattern-match expression.

Reply:
Yes, I meant point 4. But isn't `ecndec` not already described in point 2?
Because one can't see these other `encdec` definitions in the screenshot, I think it is important to mention/explain the scattered mapping or cut it for brevity.